rfcs/82-transform2d.md

```diff
@@ -0,0 +1,137 @@
+
+#### Poor Ergonomics
+
+In order to avoid issues caused by the above overloading any system that interacts with an entities translation will often store the entities Z, copy a `Vec2` of the translation using either xy() or truncate(), perform any necessary operations, then re-extend the `Vec2` back to a `Vec3` with the stored Z before reassigning the value to the Transforms translation. The effect of this is that a 2D games code is littered with truncate() and extend(). This is then further exasperated when doing operations like Raycasting where maintaining the existing Z value will give incorrect results (including Z distance in ray calculations). All in all, this is a poor user experience.
+
+### The Problem with Rotation
+
+Rotation in bevy is stored as a `Quat` in the Transform component. Again, this is fine for 3D games, and is usable for 2D games, but has poor ergonomics. Rotation in a 2D game in Bevy is always counterclockwise around the Z axis. As with using `Vec3`'s for Translation this causes a lot of boilerplate, extracting the value of a quaternions rotation around Z in radians, performing calculations on that value, then creating a new `Quat` rotating around Z before reassigning. Much as with Translation, this is a poor user experience.
+
+## User-facing explanation
+
+Transform2D represents a 2D entities position, rotation, and scale in world space. It is defined as follows:
+
+```rust
+pub struct Transform2d {
+  position: Vec2,
+  rotation: f32 //RFC Note: We could use Rotation2d here
+  scale: Vec2
+}
+```
+
+If you're making a 2D game and you want your entity to appear in the world at a given location it must have a `Transform2D` component. `Position` represents your entities location in the world (actually it's position on the 'xy-plane'). `Rotation` represents your entities anti-clockwise rotation in radians (around the Z-axis). `Scale` represents the entities scale in the world (like `Position` this is actually the entities scale on the 'xy-plane').
+
+Many entities in a 2D game will have a sprite, a mesh, or some other graphics attached to them. By default, all these entities are drawn on the same 'layer'. When entities are drawn the last entity drawn appears on top of thosed drawn before it in the same location. Developers can control which layer an entity is on by changing that entities `DrawLayer`:
+
+```rust
+pub struct DrawLayer(pub f32);
+```
+
+As you can see `DrawLayer` is just a wrapper around an `f32`. Entities with a higher value for their `DrawLayer` will be drawn on top of those with lower values.
```

Comment:
So this is one of the approaches that's been tossed around for z ordering, and personally I'm not a huge fan.

There are a few issues.
1. Floating points are not totally ordered, and by exposing them to users for ordering we invite them to make mistakes (rather than just using them for ordering internally, where we can ensure correctness). It's not clear to me what should happen if you have a NAN layer, or if negative draw layers should be allowed, or what the ordering properties for negative vs positive zero should be. Negative and positive infinity do seem fine, and I'll note that there are ways to make floats `Ord` (but they are not free).
2. You can exhaust the precision of the space; inserting a layer between two other layers can fail, require moving an arbitrary number of layers to make room. I'm not sure how common this will be in practice, but I suspect with enough nesting it will be something people run into repeatedly.
3. It seems like to determine z position, you sum the layers of all the ancestors and add it to the layer of the entity. This probably means storing a `PropigatedLayer(f32)` component, which seems like it would add complexity. "Layer" also dosn't seem like the right term to describe this; perhaps "z-offset"?

Reply:
I thought about this after I wrote the RFC, but honestly, I would prefer this to be a u32 rather than an f32.

Also yes, that's my suggestion, summing all the layers of the parents values and adding it the childs.